Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,221 @@
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftDiagnostics
+import SwiftParser
+import SwiftParserDiagnostics
+import SwiftSyntax
+import XCTest
+import _SwiftSyntaxTestSupport
+
+/// A typealias representing a location marker.
+///
+/// This string serves to pinpoint the exact location of a particular token in the Abstract Syntax Tree (AST).
+/// Once the token location is identified, it can be leveraged for various test-specific operations such as inserting diagnostics, notes, or fix-its,
+/// or for closer examination of the syntax tree.
+///
+/// Markers are instrumental in writing unit tests that require precise location data. They are commonly represented using emojis like 1️⃣, 2️⃣, 3️⃣, etc., to improve readability.
+///
+/// ### Example
+///
+/// In the following test code snippet, the emojis 1️⃣ and 2️⃣ are used as location markers:
+///
+/// ```swift
+/// func foo() -> Int {
+///   if 1️⃣1 != 0 2️⃣{
+///     return 0
+///   }
+///   return 1
+/// }
+/// ```
+typealias LocationMarker = String
+
+/// Represents a descriptor for constructing a diagnostic in testing.
+struct DiagnosticDescriptor {
+  /// The marker pointing to location in source code.
+  let locationMarker: LocationMarker
+
+  /// The ID associated with the message, used for categorizing or referencing it.
+  let id: MessageID
+
+  /// The textual content of the message to be displayed.
+  let message: String
+
+  /// The severity level of the diagnostic message.
+  let severity: DiagnosticSeverity
+
+  /// The syntax elements to be highlighted for this diagnostic message.
+  let highlight: [Syntax]  // TODO: How to create an abstract model for this?
+
+  /// Descriptors for any accompanying notes for this diagnostic message.
+  let noteDescriptors: [NoteDescriptor]
+
+  /// Descriptors for any Fix-Its that can be applied for this diagnostic message.
+  let fixIts: [FixIt]  // TODO: How to create an abstract model for this?
+
+  /// Initializes a new `DiagnosticDescriptor`.
+  ///
+  /// - Parameters:
+  ///   - locationMarker: The marker pointing to location in source code.
+  ///   - id: The message ID of the diagnostic.
+  ///   - message: The textual message to display for the diagnostic.
+  ///   - severity: The severity level of the diagnostic. Default is `.error`.
+  ///   - highlight: The syntax elements to be highlighted. Default is an empty array.
+  ///   - noteDescriptors: An array of note descriptors for additional context. Default is an empty array.
+  ///   - fixIts: An array of Fix-It descriptors for quick fixes. Default is an empty array.
+  init(
+    locationMarker: LocationMarker,
+    id: MessageID,
```

Comment:
How about defaulting this to `.test`? We don’t really care about the ID in these tests.